google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadChannel.java

```diff
@@ -101,27 +101,28 @@ public int read(ByteBuffer dst) throws IOException {
       checkOpen();
       int amt;
       int retries = 0;
-      int maxRetries = 3;
+      int maxRetries = Math.max(3, maxChannelReopens);
       dst.mark();
       while (true) {
         try {
           dst.reset();
           amt = channel.read(dst);
           break;
         } catch (StorageException exs) {
-          if (exs.getMessage().contains("Connection closed prematurely") && reopens < maxChannelReopens) {
-            // this error isn't marked as retryable since the channel is closed;
-            // but here at this higher level we can retry it.
+          if (reopens < maxChannelReopens && (isReopenable(exs) || isReopenable(exs.getCause()))) {
```

Comment:
Thanks -- as a nice side effect, now that you're iterating down the chain in `isReopenable()`, you shouldn't need to call both `isReopenable(exs)` and `isReopenable(exs.getCause())` here -- `isReopenable(exs)` alone should suffice.